torch/csrc/jit/argument_spec.h

```diff
@@ -0,0 +1,180 @@
+#include <Python.h>
+#include <iostream>
+#include <vector>
+#include "torch/csrc/autograd/variable.h"
+#include "torch/csrc/utils/hash.h"
+
+namespace torch { namespace jit {
+
+// GraphExecutor creates specializations of Graphs for different dimensionalitities
+// and types of inputs.
+
+// ArgumentSpec represents one particular specialization.
+// It is designed so that it can be created, hashed, and compared quickly
+// since it is used along the hot-path of the JIT to check if the code
+// we have created is valid for the given inputs.
+
+// TensorInfoPOD is only used internally in ArgumentSpec
+// API users should use TensorInfo
+struct TensorInfoPOD {
+  // total size is 64-bit
+  unsigned int type : 8;
+  unsigned int defined : 1;
+  unsigned int requires_grad : 1;
+  int device : 22;
+  uint32_t total_dims; // all TensorInfoPODs are in ArgumentSpec's tensor_info() array.
+                       // total_dims is the total number of dimensions seen so far
+                       // in all previous members of tensor_info(), including this tensor
+                       // 2*total_dims becomes the offset into the sizes_strides list
+                       // for the _next_ tensor in the tensor_info array
+                       // for tensor 0, the offset is always 0
+};
+static_assert(sizeof(TensorInfoPOD) == sizeof(int64_t),
+  "TensorInfoPOD must be 64-bit struct for ArgumentSpec encoding to work");
+
+struct TensorInfo;
+
+struct ArgumentSpec {
+  ArgumentSpec(bool nograd, ArrayRef<autograd::Variable> tensors)
```

Comment:
Can we not use `nograd`? We need to use double negation later. It's often much clearer to always use positive flags.